### PR TITLE
PDF Engine Selection

### PR DESCRIPTION
**Pull request by Dillon-Brown**
## Description

- Initial Framework for PDF Engine selection (At the moment we only have the single Engine).
- Config generation will be handled on upgrade.

## How To Test This

1. See new admin panel option "PDF Settings"
2. Check that saving correctly generates a config value: ```$sugar_config['pdf']['defaultEngine'] = 'MPDFEngine';```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [X] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: lib/PDF/PDFConfigurator.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ * Class PDFConfigurator
+ * @package SuiteCRM\PDF
+ */
+class PDFConfigurator
+{
+    /** @var Configurator */
+    private $configurator;
+
+    /**
+     * SearchConfigurator constructor.
```
Comment (pankaj0882) [inline lib/PDF/PDFConfigurator.php:61]: name of the class

File: lib/PDF/PDFConfigurator.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /**
+     * @param string $engine
+     * @return $this
+     */
+    public function setEngine(string $engine): PDFConfigurator
+    {
+        if (empty($engine)) {
+            throw new InvalidArgumentException('$engine cannot be empty');
```
Comment (pankaj0882) [inline lib/PDF/PDFConfigurator.php:91]: the exception will display as
' ' cannot be empty ? 
So, better to replace the variable with a text; say, Engine can not be empty Or, Engine name can not be empty.

File: modules/Administration/PDF/MVC/View.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        $errors = [];
+        $this->smarty->assign('MOD', $mod_strings);
+        $this->smarty->assign('APP', $app_strings);
+        $this->smarty->assign('APP_LIST', $app_list_strings);
+        $this->smarty->assign('LANGUAGES', get_languages());
+        $this->smarty->assign('JAVASCRIPT', get_set_focus_js());
+        $this->smarty->assign('error', $errors);
+        $this->smarty->assign('BUTTONS', $this->getButtons());
+
+        if (!empty($sugar_config['pdf'])) {
```
Comment (pankaj0882) [inline modules/Administration/PDF/MVC/View.php:74]: not sure, if this wil be `if (empty($sugar_config['pdf'])) {`

File: modules/Administration/PDF/MVC/View.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            LoggerManager::getLogger()->warn('Configuration does not contains default PDF settings.');
+        }
+        $pdfSettings = $sugar_config['pdf'] ?? null;
```
Comment (pankaj0882) [inline modules/Administration/PDF/MVC/View.php:77]: not sure, if we shall filter the null value before assignment

File: lib/PDF/PDFWrapper.php [hunk 5]
```diff
@@ -172,4 +172,16 @@ private static function getPDFConfig($key)
 
         return $sugar_config['pdf'][$key] ?? null;
     }
+
+    /**
+     * Returns the configured search controller from the sugar config.
```
Comment (pankaj0882) [inline lib/PDF/PDFWrapper.php:177]: mesage containing `search controller`

**Comment by pankaj0882** [comment]
@Dillon-Brown the changes look very good except some minor comments.

File: lib/PDF/PDFConfigurator.php (hunk 1 above)
Comment (Dillon-Brown) [reply lib/PDF/PDFConfigurator.php:61]: Pushed.

File: lib/PDF/PDFConfigurator.php (hunk 2 above)
Comment (Dillon-Brown) [reply lib/PDF/PDFConfigurator.php:91]: Agreed, pushed.

File: modules/Administration/PDF/MVC/View.php (hunk 3 above)
Comment (Dillon-Brown) [reply modules/Administration/PDF/MVC/View.php:74]: Ah true, nice spot!

File: modules/Administration/PDF/MVC/View.php (hunk 4 above)
Comment (Dillon-Brown) [reply modules/Administration/PDF/MVC/View.php:77]: Should be fine to have this as ```null```, would want this assigned to handle cases where a user has set ```$sugar_config['pdf'] = ''```.

File: lib/PDF/PDFWrapper.php (hunk 5 above)
Comment (Dillon-Brown) [reply lib/PDF/PDFWrapper.php:177]: Pushed :)

**Review by pankaj0882 — COMMENTED** [review]
Looks Good To Me :)

File: modules/Administration/PDF/Controller.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Class Controller
+ * @package SuiteCRM\Modules\Administration\PDF
+ */
+class Controller extends AbstractController
+{
+    public function __construct()
+    {
+        parent::__construct(new View());
```
Comment (clemente-raposo) [inline modules/Administration/PDF/Controller.php:58]: Hi @Dillon-Brown, don't we need to declare the `View` on the imports?

File: modules/Administration/PDF/MVC/View.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+     * Returns the cancel and save button.
+     *
+     * @return string
+     */
+    protected function getButtons(): string
+    {
+        global $mod_strings;
+        global $app_strings;
+
+        return <<<EOQ
```
Comment (clemente-raposo) [inline modules/Administration/PDF/MVC/View.php:92]: Hi @Dillon-Brown, totally optional, though if you agreed and have some time, could we add this to a separate tpl file?

File: modules/Administration/PDF/View.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    die('Not A Valid Entry Point');
+}
+
+use SuiteCRM\Modules\Administration\PDF\MVC\View as AbstractView;
+use SuiteCRM\PDF\PDFWrapper;
+
+/**
+ * Class View renders the PDF settings.
+ */
+class View extends AbstractView
```
Comment (clemente-raposo) [inline modules/Administration/PDF/View.php:52]: Hi @Dillon-Brown, the class on `modules/Administration/PDF/MVC/View.php` is also called `View`.  Would you be ok with changing the name to another name to help better understand which of the two we are using?

File: modules/Administration/PDF/Controller.php (hunk 6 above)
Comment (Dillon-Brown) [reply modules/Administration/PDF/Controller.php:58]: Both are using the same namespace so it shouldn't need to be imported.

File: modules/Administration/PDF/MVC/View.php (hunk 7 above)
Comment (Dillon-Brown) [reply modules/Administration/PDF/MVC/View.php:92]: Agreed, pushed!

File: modules/Administration/PDF/View.php (hunk 8 above)
Comment (Dillon-Brown) [reply modules/Administration/PDF/View.php]: Agreed, pushed!

**Review by clemente-raposo — DISMISSED** [review]
Changes look ok

**Comment by MacRobberts** [comment]
Assessed :+1:

**Review by MacRobberts — APPROVED** [review]

**Comment by Mac-Rae** [comment]
Assessed :+1:

**Review by Mac-Rae — APPROVED** [review]

**Review by clemente-raposo — APPROVED** [review]
rebase changes look ok

**Pull request merged**